networkx/algorithms/dag.py

```diff
@@ -32,69 +29,32 @@
     "dag_to_branching",
 ]
 
-chaini = chain.from_iterable
-
-
-def descendants(G, source):
-    """Returns all nodes reachable from `source` in `G`.
-
-    Parameters
-    ----------
-    G : NetworkX Graph
-    source : node in `G`
-
-    Returns
-    -------
-    set()
-        The descendants of `source` in `G`
-
-    Raises
-    ------
-    NetworkXError
-        If node `source` is not in `G`.
-
-    Examples
-    --------
-    >>> DG = nx.path_graph(5, create_using=nx.DiGraph)
-    >>> sorted(list(nx.descendants(DG, 2)))
-    [3, 4]
+_deprecated_names = [
+    "descendants",
+    "ancestors",
+    "transitive_closure",
+]
 
-    See also
-    --------
-    ancestors
-    """
-    return {child for parent, child in nx.bfs_edges(G, source)}
 
+def __getattr__(name):
+    f = getattr(nx, name)
```

Comment:
At face value this seems like an elegant way to grab the definitions from their new modules, but this has dangerous consequences! This basically exposes the entire `nx` namespace from the `dag` module, which is definitely not what we want. For example, with this code I think you'd be able to do something like:

```python
from networkx.algorithms.dag import complement  # Note: complement is *not* defined in dag.py
```

and it would work, though it definitely shouldn't.

I think the fix is to get rid of this line, and be explicit about the imports, maybe using `importlib` e.g. something like

```python
if name in _deprecated_names:
    ...
    import importlib
    f = getattr(importlib.import_module("networkx.algorithms.traversal.reachability"), name)
```